vumi/transports/smpp/tests/test_protocol.py

```diff
@@ -121,100 +123,97 @@ def test_on_connection_made(self):
         self.assertTrue(protocol.enquire_link_call.running)
 
     @inlineCallbacks
     def test_handle_unbind(self):
-        transport, protocol = yield self.setup_bind()
-        protocol.dataReceived(Unbind(sequence_number=0).get_bin())
-        [pdu] = yield wait_for_pdus(transport, 1)
-        self.assertCommand(pdu, 'unbind_resp',
-                           sequence_number=0, status='ESME_ROK')
+        protocol = yield self.get_protocol()
+        yield self.fake_smsc.bind()
+        self.assertEqual(protocol.state, EsmeProtocol.BOUND_STATE_TRX)
+        self.fake_smsc.send_pdu(Unbind(0))
+        pdu = yield self.fake_smsc.await_pdu()
+        self.assertCommand(
+            pdu, 'unbind_resp', sequence_number=0, status='ESME_ROK')
+        # We don't change state here.
+        self.assertEqual(protocol.state, EsmeProtocol.BOUND_STATE_TRX)
 
     @inlineCallbacks
     def test_on_submit_sm_resp(self):
+        protocol = yield self.get_protocol()
+        yield self.fake_smsc.bind()
         calls = []
-        self.patch(EsmeProtocol, 'on_submit_sm_resp',
-                   lambda p, *a: calls.append(a))
-        transport, protocol = yield self.setup_bind()
-        pdu = SubmitSMResp(sequence_number=0, message_id='foo')
-        protocol.dataReceived(pdu.get_bin())
+        protocol.on_submit_sm_resp = lambda *a: calls.append(a)
+        yield self.fake_smsc.send_pdu(SubmitSMResp(0, message_id='foo'))
         self.assertEqual(calls, [(0, 'foo', 'ESME_ROK')])
 
     @inlineCallbacks
     def test_deliver_sm(self):
         calls = []
-        self.patch(EsmeProtocol, 'handle_deliver_sm',
-                   lambda p, pdu: succeed(calls.append(pdu)))
-        transport, protocol = yield self.setup_bind()
-        pdu = DeliverSM(
-            sequence_number=0, message_id='foo', short_message='bar')
-        protocol.dataReceived(pdu.get_bin())
+        protocol = yield self.get_protocol()
+        protocol.handle_deliver_sm = lambda pdu: succeed(calls.append(pdu))
+        yield self.fake_smsc.bind()
+        yield self.fake_smsc.send_pdu(
+            DeliverSM(0, message_id='foo', short_message='bar'))
         [deliver_sm] = calls
         self.assertCommand(deliver_sm, 'deliver_sm', sequence_number=0)
 
     @inlineCallbacks
     def test_deliver_sm_fail(self):
-        transport, protocol = yield self.setup_bind()
-        pdu = DeliverSM(
+        yield self.get_protocol()
+        yield self.fake_smsc.send_pdu(DeliverSM(
             sequence_number=0, message_id='foo', data_coding=4,
-            short_message='string with unknown data coding')
-        protocol.dataReceived(pdu.get_bin())
-        [deliver_sm_resp] = yield wait_for_pdus(transport, 1)
+            short_message='string with unknown data coding'))
+        deliver_sm_resp = yield self.fake_smsc.await_pdu()
         self.assertCommand(
             deliver_sm_resp, 'deliver_sm_resp', sequence_number=0,
             status='ESME_RDELIVERYFAILURE')
```

Comment:
`status='ESME_ROK'` is success, anything else is failure of some kind. (The throttling statuses `ESME_RTHROTTLED` and `ESME_RMSGQFUL` are special kinds of failures that we treat differently.)